nano/node/socket.cpp

```diff
@@ -0,0 +1,334 @@
+		this_l->remote = endpoint_a;
+		callback_a (ec);
+	}));
+}
+
+void nano::socket::async_read (std::shared_ptr<std::vector<uint8_t>> buffer_a, size_t size_a, std::function<void(boost::system::error_code const &, size_t)> callback_a)
+{
+	assert (size_a <= buffer_a->size ());
+	auto this_l (shared_from_this ());
+	start_timer ();
+	boost::asio::async_read (tcp_socket, boost::asio::buffer (buffer_a->data (), size_a),
+	boost::asio::bind_executor (strand,
+	[this_l, buffer_a, callback_a](boost::system::error_code const & ec, size_t size_a) {
+		if (auto node = this_l->node.lock ())
+		{
+			node->stats.add (nano::stat::type::traffic_tcp, nano::stat::dir::in, size_a);
+			this_l->stop_timer ();
+			callback_a (ec, size_a);
+		}
+	}));
+}
+
+void nano::socket::async_write (std::shared_ptr<std::vector<uint8_t>> buffer_a, std::function<void(boost::system::error_code const &, size_t)> callback_a)
+{
+	auto this_l (shared_from_this ());
+	if (writer_concurrency == nano::socket::concurrency::multi_writer)
+	{
+		boost::asio::post (strand,
+		[buffer_a, callback_a, this_l]() {
+			bool write_in_progress = !this_l->send_queue.empty ();
+			this_l->send_queue.emplace_back (nano::socket::queue_item{ buffer_a, callback_a });
+			if (!write_in_progress)
+			{
+				this_l->write_queued_messages ();
+			}
+		});
+	}
+	else
+	{
+		start_timer ();
+		boost::asio::async_write (tcp_socket, boost::asio::buffer (buffer_a->data (), buffer_a->size ()),
+		boost::asio::bind_executor (strand,
+		[this_l, buffer_a, callback_a](boost::system::error_code const & ec, size_t size_a) {
+			if (auto node = this_l->node.lock ())
+			{
+				node->stats.add (nano::stat::type::traffic_tcp, nano::stat::dir::out, size_a);
+				this_l->stop_timer ();
+				if (callback_a)
+				{
+					callback_a (ec, size_a);
+				}
+			}
+		}));
+	}
+}
+
+void nano::socket::write_queued_messages ()
+{
+	std::weak_ptr<nano::socket> this_w (shared_from_this ());
+	auto msg (send_queue.front ());
+	start_timer ();
+	boost::asio::async_write (tcp_socket, boost::asio::buffer (msg.buffer->data (), msg.buffer->size ()),
+	boost::asio::bind_executor (strand,
+	[msg, this_w](boost::system::error_code ec, std::size_t size_a) {
+		if (auto this_l = this_w.lock ())
+		{
+			if (auto node = this_l->node.lock ())
+			{
+				node->stats.add (nano::stat::type::traffic_tcp, nano::stat::dir::out, size_a);
+			}
+			this_l->stop_timer ();
+
+			if (!this_l->closed)
+			{
+				if (msg.callback)
+				{
+					msg.callback (ec, size_a);
+				}
+
+				this_l->send_queue.pop_front ();
+				if (!ec && !this_l->send_queue.empty ())
+				{
+					this_l->write_queued_messages ();
+				}
+			}
+		}
+	}));
+}
+
+void nano::socket::start_timer ()
+{
+	if (auto node_l = node.lock ())
+	{
+		start_timer (node_l->config.tcp_io_timeout);
+	}
+}
+
+void nano::socket::start_timer (std::chrono::seconds deadline_a)
+{
+	std::weak_ptr<nano::socket> this_w (shared_from_this ());
+	boost::asio::post (strand, [this_w, deadline_a]() {
```

Comment:
The old way of starting and stopping timers was to queue actions on the alarm class which had issues when there were a high number of start/stops queued and memory usage would balloon, is this an issue here as well with queueing items on a strand?
This was replaced with an atomic ticket system where we could detect if a timer had been stopped without a memory allocation.

Reply:
Good point about posting through a strand, reverting to using an atomic ticket system.

Reply:
I'll see if I can adapt the atomic ticket system to work with concurrent writers.